### PR TITLE
CVS-84589 No copy demultiplexer

### PR DESCRIPTION
**Pull request by dkalinowski**
This PR changes session results handling to work with `TensorWithSource` class instead of raw `ov::Tensor`. This way we can collect demultiplied results together with its source tensor and keep it for the lifetime of sharded tensor. This creates posibility to not copy, the sharded tensors contain reference to source tensor data.

### CODE REVIEW

File: src/node.cpp [hunk 1]
```diff
@@ -131,11 +131,14 @@ Status Node::setInputs(const Node& dependency, TensorMap& inputs, NodeSessionMet
             dependency.getName(),
             current_node_input_name,
             dependency_output_name);
-        auto status = nodeSession->setInput(current_node_input_name, it->second, shardId);
+        auto status = nodeSession->setInput(current_node_input_name, it->second.getActualTensor(), shardId);
```
Comment (atobiszei) [inline src/node.cpp:134]: Wouldn't it make sense to instead of asking of original tensor for source just pass this wrapper into InputHandler?

Comment (dkalinowski) [reply src/node.cpp:134]: If we saved the wrapper in InputHandler we would need to pay for creating empty ov::Tensor() (source tensor) even if there is no demultiplexer in pipeline

Comment (dkalinowski) [reply src/node.cpp]: Adjusted after sync.

**Review by michalkulakowski — APPROVED** [review]

File: src/test/demultiplexer_node_test.cpp [hunk 2]
```diff
@@ -186,3 +186,105 @@ TEST(DemultiplexerTest, ShardsShareDataWithSourceTensor) {
...
+    auto dl_gather = std::make_unique<MockDLNode>(
+        std::string("dummy_node_2"),
+        std::string("dummy_model"),
+        1, manager,
+        std::unordered_map<std::string, std::string>{{"b", "b"}},
+        std::nullopt,
+        std::set<std::string>{"dummy_node_1"});
+
+    dl_demulti->addDependant(*dl_gather);
+    dl_gather->addDependency(*dl_demulti, Aliases{{"b", "b"}});
```
Comment (atobiszei) [inline src/test/demultiplexer_node_test.cpp:238]: "b" -> DUMMY_MODEL_INPUT_NAME

File: src/test/demultiplexer_node_test.cpp [hunk 3]
```diff
@@ -186,3 +186,105 @@ TEST(DemultiplexerTest, ShardsShareDataWithSourceTensor) {
...
+
+    // Prepare source tensors
+    std::vector<float> data{3.2, 5.9};
+    std::vector<std::shared_ptr<ov::Tensor>> tensors;
+    for (int i = 0; i < demultiplyCount; i++) {
+        tensors.push_back(std::make_shared<ov::Tensor>(ov::element::f32, ov::Shape{1, size}));
+        std::memcpy(tensors[i]->data(), data.data(), sizeof(float) * size);
+    }
+
+    // Prepare session results and pass to ::setInputs method - source and actual tensors.
```
Comment (atobiszei) [inline src/test/demultiplexer_node_test.cpp:250]: Add comment that in this section we imitate demultiplication

**Review by atobiszei — APPROVED** [review]

File: src/test/demultiplexer_node_test.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/test/demultiplexer_node_test.cpp]: Done

File: src/test/demultiplexer_node_test.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/test/demultiplexer_node_test.cpp:250]: Done

**Pull request merged**